cmd/xdpcap/flags.go

```diff
@@ -28,15 +32,136 @@ func (a *actionsFlag) Set(val string) error {
 	return nil
 }
 
-func (a *actionsFlag) String() string {
+func (a actionsFlag) String() string {
 	strs := []string{}
-	for _, action := range *a {
+	for _, action := range a {
 		strs = append(strs, action.String())
 	}
 
 	return strings.Join(strs, ",")
 }
 
+type linkTypeFlag layers.LinkType
+
+func (l *linkTypeFlag) Set(val string) error {
+	val = strings.TrimSpace(strings.ToLower(val))
+
+	for _, link := range linkTypes() {
+		if val == strings.ToLower(link.String()) {
+			*l = linkTypeFlag(link)
+			return nil
+		}
+	}
+
+	// Accept links we don't know about using their numeric value
+	// Base 0 to allow for hex
+	unknownLink, err := strconv.ParseInt(val, 0, 8)
+	if err != nil {
+		return errors.Errorf("unknown linktype %s", val)
+	}
+
+	*l = linkTypeFlag(unknownLink)
+	return nil
+}
+
+func (l linkTypeFlag) String() string {
+	return strings.ToLower(layers.LinkType(l).String())
+}
+
+// All valid LinkTypes
+func linkTypes() []linkTypeFlag {
+	links := []linkTypeFlag{}
+
+	for link, meta := range layers.LinkTypeMetadata {
+		if meta.Name == "UnknownLinkType" {
+			continue
+		}
+
+		links = append(links, linkTypeFlag(link))
+	}
+
+	return links
+}
+
+var bpfRegex = regexp.MustCompile(`^\d+(?:,\d+ \d+ \d+ \d+)+,?$`)
+
+func parseFilter(expr string, linkType layers.LinkType) ([]bpf.Instruction, error) {
+	expr = strings.TrimSpace(expr)
+
+	if bpfRegex.MatchString(expr) {
+		return parsecBPF(expr)
+	}
+
+	return tcpdumpExprToBPF(expr, linkType)
+}
+
+// parsecBPF parses a list of cBPF 4 tuple instructions
+func parsecBPF(bpfStr string) ([]bpf.Instruction, error) {
+	cbpf := strings.Split(strings.TrimSuffix(bpfStr, ","), ",")
+
+	if len(cbpf) < 1 {
+		return nil, errors.Errorf("unable to split cBPF length & instructions")
+	}
+
+	insCount, err := strconv.Atoi(cbpf[0])
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to parse cBPF length")
+	}
+
+	insns := cbpf[1:]
+	if len(insns) != insCount {
+		return nil, errors.Errorf("declared cBPF instruction length %d doesn't match actual %d", insCount, len(insns))
+	}
+
+	cbpfInsns := make([]bpf.Instruction, len(insns))
+	for i, insnStr := range insns {
+		cbpfInsns[i], err = parseInstruction(insnStr)
+		if err != nil {
+			return nil, errors.Wrapf(err, "instruction %d", i)
+		}
+	}
+
+	return cbpfInsns, nil
+}
+
+// parseInstruction parses a string of cBPF 4 tuple instructions, formatted as:
+// <length>,<opcode> <jt> <jf> <k>,...
```

Comment:
It doesn't parse length though?

Reply:
Oops yes, that comment is wrong. Updated :)